### PR TITLE
Use transpileModule for TS and transform for babel conditionally

### PR DESCRIPTION
**Pull request by lmiller1990**
Seems to fix https://github.com/vuejs/vue-cli/issues/5714. Tested locally with latest vue-cli with a TS + babel and no babel setup.

### CODE REVIEW

File: lib/process.js [hunk 1]
```diff
@@ -68,10 +70,19 @@ function processTemplate(template, filename, config) {
...
 
-  const babelify = transform(result.code, { filename: 'file.js' })
+  const tsconfig = getTsJestConfig(config)
 
-  return {
-    code: babelify.code
+  if (tsconfig) {
+    // they are using TypeScript.
+    const { outputText } = transpileModule(result.code, { tsconfig })
+    return { code: outputText }
```
Comment (ahnpnl) [inline lib/process.js:78]: If it is a combination of ts and Babel, I think you should do transform twice, once with typescript and once with Babel. Should it be like that ?

**Comment by ahnpnl** [comment]
One of the differences between transforming with ts vs Babel is: ts doesn’t include polyfills while Babel does.

An example is: Array.prototype.flat which is available on node 11. When transforming ts to js, ts won’t polyfill it. If chaining Babel into the transform pipe, Babel will polyfill it.

File: lib/process.js (hunk 1 above)
Comment (lmiller1990) [reply lib/process.js:78]: Hm, I am not sure in this particular case it makes any difference. Not we are transforming the code generated by the `@vue/sfc-compiler`, not an actual human, so I am not sure we need to care about polyfills. I could be wrong.

**Comment by lmiller1990** [comment]
Hm, interesting. I am not sure we need the double transform here; this library (vue-jest) doesn't have support for running in a browser right now, so everyone is running in a node + jsdom environment. Unless you are on a fairly old node version, the code generated from `@vue/compiler-sfc`  should work just fine without polyfills.

We could run it through babel to be sure, I guess - I wonder what kind of performance hit this would incur, though? For some very large test suites, this might not be ideal.

**Comment by ahnpnl** [comment]
I think it’s safe to have e2e test for different combinations

**Comment by lmiller1990** [comment]
Looks like our e2e tests are failing!

**Comment by sodatea** [comment]
Have you tried optional chaining syntax in the template with Vue CLI TS + Babel setup?

**Comment by sodatea** [comment]
(with Node.js v12)

**Comment by lmiller1990** [comment]
I have not tried that, is there any specific thing I should be trying to chain? I have actually never used the optional chaining feature in Vue CLI, what chaining would you expect to potentially break things?

**Comment by sodatea** [comment]
IIRC, in that setup, TS preserves the `?.` syntax, and Babel will then transpile it. So skipping Babel transpilation may break it in older Node versions. (Node 14 supports this syntax natively so we need to test in Node 12)

**Comment by lmiller1990** [comment]
Oh I thought you were referring to how you can chain the webpack configuration in `vue.config.js`. I understand now. I will test that.

**Comment by lmiller1990** [comment]
@sodatea I tried both with the null operator and (somewhat surprisingly) it worked fine.

I tried:
- [x] ts + babel
- [x] ts only
- [x] babel only

Here is the test I used:

```vue
<template>
  <div class="hello">
    <h1>{{ msg?.english }}</h1>
  </div>
</template>

<script lang="ts">
import { defineComponent } from 'vue';

export default defineComponent({
  name: 'HelloWorld',
  props: {
    msg: Object
  },
});
</script>
```

```ts
import { shallowMount } from '@vue/test-utils'
import HelloWorld from '@/components/HelloWorld.vue'

describe('HelloWorld.vue', () => {
  it('renders props.msg when passed', () => {
    const english = 'new message'
    const wrapper = shallowMount(HelloWorld, {
      props: { msg: { english } }
    })
    expect(wrapper.text()).toMatch(english)
  })
})
```

I am surprised babel knew about the null operator. 

Can you think of anything else I should test out?

**Comment by sodatea** [comment]
Good to know that!
I think it's good to go now.

**Pull request merged**